### DISCUSSION TITLE
How to execute several processes sequentially, with channels in DSL2?

### DISCUSSION THREAD

**Discussion by satish15625** [nextflow-io/nextflow#3415] [Q&A] [answered]
[process.txt](https://github.com/nextflow-io/nextflow/files/10055387/process.txt)

**Comment by mribeirodantas** [comment, marked as answer]
Adding `maxForks 1` as a process directive will make your tasks run one by one, with no implicit parallelization. If you run the code below, you'll probably see 3 tasks concluding simultaneosly, then 6, 9 and 10. If you set `maxForks 1` (now it's `maxForks 3`), you'll see 1, 2, 3, 4, 5, 6, 7, 8, 9 and 10 because it is run sequentially.

```
process FOO {
  maxForks 3
  input:
    val number
  script:
  """
  echo $number
  sleep 1
  """
}

workflow {
  Channel
    .of(1..10)
  | FOO
}
```

**Reply by mribeirodantas** [reply]
An update to this answer, as for the comment by Paolo below. You can make Nextflow request that tasks of a process are done one by one with `maxForks 1`, but there is no guarantee that they will be done one by one **in the order that the channel dispatches them**. The `fair` directive to be available soon will fix that 🥳

**Comment by pditommaso** [comment]
Sequential execution is not possible as of version 22.11.0-edge. However upcoming version of nextflow will introduce a new `fair` directive. When enabling it the process output is guaranteed to be produced in the same sequence as the inputs where received. For example 

```
process sayHello {
  fair true
  input: val(x)
  output: stdout 
  """
  echo -n $x
  """
} 

workflow {
  def ch1 = channel.of('Hello','Ciao','Hola','Boujour')
  def ch2 = channel.of('world!','mondo!','mundo!','le monde!')

  ch1 | sayHello | merge(ch2) | view((x,y) -> "$x $y")
}
```  


it will print 

```
Hello world!
Ciao mondo!
Hola mundo!
Boujour le monde!
```
